File: cli/src/wsc/_url.py
"""URL helpers — redaction and normalization.

Lives in its own module so both `wsc.audit` and `wsc.providers.*` can import
it without pulling in either side's surface area.
"""

from __future__ import annotations

from typing import Iterable
from urllib.parse import parse_qsl, urlencode, urlsplit, urlunsplit

# Query-string keys whose values must never reach disk or stdout. Lowercased.
_SECRET_KEYS = frozenset(
    {
        "token",
        "tokens",
        "access_token",
        "id_token",
        "refresh_token",
        "auth",
        "authkey",
        "apikey",
        "api_key",
        "api-key",
        "x-api-key",
        "key",
        "secret",
        "client_secret",
        "sig",
        "signature",
        "hmac",
        "session",
        "sessionid",
        "password",
        "passwd",
        "pwd",
    }
)

# Tracking parameters we drop (pure noise, also good for dedup).
_TRACKING_KEYS = frozenset(
    {
        "utm_source",
        "utm_medium",
        "utm_campaign",
        "utm_term",
        "utm_content",
        "utm_id",
        "gclid",
        "fbclid",
        "mc_cid",
        "mc_eid",
        "ref",
        "_hsenc",
        "_hsmi",
    }
)


def redact_url(url: str) -> str:
    """Replace secret query values with ``***`` while leaving normal params intact.

    Returns the input unchanged for empty / non-URL strings rather than raising
    so call sites can apply this defensively.
    """

    if not url or "://" not in url:
        return url
    parts = urlsplit(url)
    if not parts.query:
        return url
    new_qs = []
    for k, v in parse_qsl(parts.query, keep_blank_values=True):
        if k.lower() in _SECRET_KEYS:
            new_qs.append((k, "***"))
        else:
            new_qs.append((k, v))
    return urlunsplit(
        (parts.scheme, parts.netloc, parts.path, urlencode(new_qs), parts.fragment)
    )


def normalize_url(url: str, *, drop_tracking: bool = True) -> str:
    """Canonical form for dedup keys.

    Lower-case the host, strip default ports, drop the fragment, drop tracking
    query params, sort the remaining params for stability. Path case is left
    alone (paths are case-sensitive on most servers).
    """

    if not url or "://" not in url:
        return url
    parts = urlsplit(url)
    netloc = parts.hostname.lower() if parts.hostname else ""
    if parts.port and not _is_default_port(parts.scheme, parts.port):
        netloc = f"{netloc}:{parts.port}"
    if parts.username or parts.password:
        # Drop credentials embedded in the URL — they are always sensitive.
        pass

    qs_pairs = [
        (k, v)
        for k, v in parse_qsl(parts.query, keep_blank_values=True)
        if not (drop_tracking and k.lower() in _TRACKING_KEYS)
    ]
    qs_pairs.sort()
    path = parts.path or "/"
    return urlunsplit((parts.scheme.lower(), netloc, path, urlencode(qs_pairs), ""))


def _is_default_port(scheme: str, port: int) -> bool:
    return (scheme == "http" and port == 80) or (scheme == "https" and port == 443)


def redact_each(urls: Iterable[str]) -> list[str]:
    return [redact_url(u) for u in urls]

File: src/_url.ts
/**
 * URL helpers — redaction and normalization.
 *
 * Lives in its own module so audit.ts and providers/* can import it
 * without dragging in either side's surface area.
 */

const SECRET_KEYS = new Set(
  [
    "token",
    "tokens",
    "access_token",
    "id_token",
    "refresh_token",
    "auth",
    "authkey",
    "apikey",
    "api_key",
    "api-key",
    "x-api-key",
    "key",
    "secret",
    "client_secret",
    "sig",
    "signature",
    "hmac",
    "session",
    "sessionid",
    "password",
    "passwd",
    "pwd",
  ].map((s) => s.toLowerCase()),
);

const TRACKING_KEYS = new Set(
  [
    "utm_source",
    "utm_medium",
    "utm_campaign",
    "utm_term",
    "utm_content",
    "utm_id",
    "gclid",
    "fbclid",
    "mc_cid",
    "mc_eid",
    "ref",
    "_hsenc",
    "_hsmi",
  ].map((s) => s.toLowerCase()),
);

/**
 * Replace secret query values with `***` while leaving normal params intact.
 * Returns the input unchanged for empty / non-URL strings rather than throwing
 * so call sites can apply this defensively before logging.
 */
export function redactUrl(url: string): string {
  if (!url || !url.includes("://")) return url;
  let u: URL;
  try {
    u = new URL(url);
  } catch {
    return url;
  }
  if (!u.search) return url;
  let mutated = false;
  for (const k of Array.from(u.searchParams.keys())) {
    if (SECRET_KEYS.has(k.toLowerCase())) {
      u.searchParams.set(k, "***");
      mutated = true;
    }
  }
  return mutated ? u.toString() : url;
}

/**
 * Canonical form for dedup keys: lowercase host, default-port stripped (URL
 * does this automatically), no fragment, tracking query params dropped,
 * remaining params sorted for stability. Path case is preserved (paths are
 * case-sensitive on most servers).
 */
export function normalizeUrl(
  url: string,
  opts: { dropTracking?: boolean } = {},
): string {
  const dropTracking = opts.dropTracking ?? true;
  if (!url || !url.includes("://")) return url;
  let u: URL;
  try {
    u = new URL(url);
  } catch {
    return url;
  }
  u.hash = "";
  u.username = "";
  u.password = "";
  if (dropTracking) {
    for (const k of Array.from(u.searchParams.keys())) {
      if (TRACKING_KEYS.has(k.toLowerCase())) u.searchParams.delete(k);
    }
  }
  // Stable sort of remaining params.
  const entries = Array.from(u.searchParams.entries()).sort(([a], [b]) =>
    a < b ? -1 : a > b ? 1 : 0,
  );
  // Rebuild search.
  u.search = "";
  for (const [k, v] of entries) u.searchParams.append(k, v);
  // Force scheme + host lowercase (URL already does this, but be explicit).
  return `${u.protocol.toLowerCase()}//${u.host.toLowerCase()}${u.pathname}${u.search}`;
}

export function redactEach(urls: readonly string[]): string[] {
  return urls.map((u) => redactUrl(u));
}
